pkg/serializer/func_queue.go

```diff
@@ -3,100 +3,52 @@
-	f        func() error
-	waitFunc WaitFunc
-}
-
 type FunctionQueue struct {
-	queue  chan queuedFunction
+	queue  chan func() error
 	stopCh chan struct{}
+	err    error
 }
 
 // NewFunctionQueue returns a FunctionQueue that will be used to execute
 // functions in the same order they are enqueued.
 func NewFunctionQueue(queueSize uint) *FunctionQueue {
 	fq := &FunctionQueue{
-		queue:  make(chan queuedFunction, queueSize),
+		queue:  make(chan func() error),
 		stopCh: make(chan struct{}),
 	}
 	go fq.run()
 	return fq
 }
 
-// run starts the FunctionQueue internal worker. It will be stopped once
-// `stopCh` is closed or receives a value.
+// run the queue's internal worker. Returns when stopCh is closed or when
+// a function has been dequeued and executed. Closes stopCh after invoking
+// a function.
 func (fq *FunctionQueue) run() {
-	for {
-		select {
-		case <-fq.stopCh:
-			return
-		case f := <-fq.queue:
-			retries := 0
-			for {
-				select {
-				case <-fq.stopCh:
-					return
-				default:
-				}
-				retries++
-				if err := f.f(); err != nil {
-					if !f.waitFunc(retries) {
-						break
-					}
-				} else {
-					break
-				}
-			}
-		}
+	select {
+	case f := <-fq.queue:
+		fq.err = f()
+		// Unblock all callers to Wait().
+		close(fq.stopCh)
+	case <-fq.stopCh:
+		return
 	}
 }
 
-// Stop stops the function queue from processing the functions on the queue.
-// If there are functions in the queue waiting for them to be processed, they
-// won't be executed.
-func (fq *FunctionQueue) Stop() {
-	close(fq.stopCh)
+// Wait for the queue to be stopped.
+//
+// Returns any error returned by an enqueued function.
+func (fq *FunctionQueue) Wait() error {
+	<-fq.stopCh
+	return fq.err
 }
 
-// Wait until the FunctionQueue is stopped, or the specified context deadline
-// expires. Returns the error from the context, or nil if the FunctionQueue
-// was completed before the context deadline.
-func (fq *FunctionQueue) Wait(ctx context.Context) error {
+// Enqueue f to the queue. Blocks if the queue is full.
+// Returns immediately without enqueueing if the queue has been closed.
+func (fq *FunctionQueue) Enqueue(f func() error) {
 	select {
 	case <-fq.stopCh:
-	case <-ctx.Done():
-	}
-	if err := ctx.Err(); err != nil {
-		return fmt.Errorf("serializer %s", err)
+		return
+	default:
 	}
-	return nil
-}
-
-// Enqueue enqueues the receiving function `f` to be executed by the function
-// queue. Depending on the size of the function queue and the amount
-// of functions queued, this function can block until the function queue
-// is ready to receive more requests.
-// If `f` returns an error, `waitFunc` will be executed and, depending on the
-// return value of `waitFunc`, `f` will be executed again or not.
-// The return value of `f` will not be logged and it's up to the caller to log
-// it properly.
-func (fq *FunctionQueue) Enqueue(f func() error, waitFunc WaitFunc) {
-	fq.queue <- queuedFunction{f: f, waitFunc: waitFunc}
+	fq.queue <- f
```

Comment:
If two goroutines (A,B) call `Enqueue()` within quick succession and the function takes a reasonably long time, then A could send the function to the goroutine launched by `NewFunctionQueue()` and B will block forever. The backing goroutine would read the function from A and execute it, close the `fq.stopCh` and return without reading again the function from B.

I don't think this library is necessarily intended to be threadsafe, but in that case we should probably either document that callers should never call `Enqueue()` without ensuring exclusion, or modify this structure such that concurrent calls to Enqueue don't result in blocking the latter goroutine(s) indefinitely.